pylot/simulation/traffic_light_invasion_sensor_operator.py

```diff
@@ -0,0 +1,203 @@
+                wp_dir = wp.transform.get_forward_vector()
+                dot_ve_wp=ve_dir.x * wp_dir.x + ve_dir.y * wp_dir.y + \
+                        ve_dir.z * wp_dir.z
+
+                # Check the lane until all the "tail" has passed
+                if tail_wp.road_id == wp.road_id and \
+                        tail_wp.lane_id == wp.lane_id and dot_ve_wp > 0:
+                    # This light is red and is affecting our lane
+                    yaw_wp = wp.transform.rotation.yaw
+                    lane_width = wp.lane_width
+                    location_wp = wp.transform.location
+
+                    lft_lane_wp = self.rotate_point(
+                        carla.Vector3D(0.4 * lane_width, 0.0, location_wp.z),
+                        yaw_wp + 90)
+                    lft_lane_wp = location_wp + carla.Location(lft_lane_wp)
+                    rgt_lane_wp = self.rotate_point(
+                        carla.Vector3D(0.4 * lane_width, 0.0, location_wp.z),
+                        yaw_wp - 90)
+                    rgt_lane_wp = location_wp + carla.Location(rgt_lane_wp)
+
+                    # Is the vehicle traversing the stop line?
+                    if self.is_vehicle_crossing_line(
+                        (tail_close_pt, tail_far_pt),
+                        (lft_lane_wp, rgt_lane_wp)):
+                        location = traffic_light.get_transform().location
+                        message = TrafficInfractionMessage(
+                            TrafficInfractionType.RED_LIGHT_INVASION,
+                            Location.from_carla_location(location),
+                            msg.timestamp)
+                        self._traffic_light_invasion_stream.send(message)
+                        self._last_red_light_id = traffic_light.id
+                        break
+
+    def run(self):
+        # Read the vehicle ID from the vehicle ID stream.
+        vehicle_id_msg = self._vehicle_id_stream.read()
+        vehicle_id = vehicle_id_msg.data
+        self._logger.debug("@[{}]: Received Vehicle ID: {}".format(
+            vehicle_id_msg.timestamp, vehicle_id))
+
+        # Connect to the world.
+        _, world = get_world(self._flags.carla_host, self._flags.carla_port,
+                             self._flags.carla_timeout)
+        if world is None:
+            raise ValueError("There was an issue connecting to the simulator.")
+        self._world, self._map = world, world.get_map()
+
+        # Retrieve all the traffic lights from the world.
+        while len(world.get_actors()) == 0:
+            time.sleep(1)
+        for _actor in world.get_actors():
+            if 'traffic_light' in _actor.type_id:
+                center, waypoints = self.get_traffic_light_waypoints(_actor)
+                self._traffic_lights.append((_actor, center, waypoints))
+
+        # Retrieve the vehicle.
+        self._vehicle = get_vehicle_handle(world, vehicle_id)
+
+    def rotate_point(self, point, angle):
```

Comment:
Can you please move this method to the Vector3D class?

Reply:
Makes sense, will do!